RDRP-215 - Implement Mean of Ratios Methods

dpurches opened the pull request:
## Pull Request submission

Add functionality to calculate MoR links and apply these to carried forwards variables.
#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #


#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)


#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

dpurches committed:
Add code to create linked_df.

dpurches committed:
Add functionality to calculate MoR.

dpurches committed:
Add comments and make sure _prev columns are retained for target vars.

dpurches committed:
Flake8 Fixes.

AnneONS committed:
Merge remote-tracking branch 'origin' into RDRP-215

AnneONS commented on src/imputation/tmi_imputation.py:
```diff
@@ -111,12 +112,13 @@ def create_imp_class_col(
+    if use_cellno:
+        fil_df = filter_by_column_content(df, "cellnumber", [817])
+        # Create class col with concatenation + 817
+        fil_df[class_name] = fil_df[class_name] + "_817"
```
good way of implementing different class structure for MoR

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +121,108 @@ def carry_forwards(df, backdata, impute_vars):
     # Drop merge related columns
-    to_drop = [column for column in df.columns if column.endswith("_prev")]
+    to_drop = [column for column in df.columns
+               if (column.endswith("_prev")) & (column not in impute_vars)]
```
good decision to keep the _prev for the impute_vars

dpurches replied:
As you've said in the main comment, can't seem to see them in the imputation qa file - going to track down why this happening now.

dpurches replied:
Have updated this now - weren't selecting the columns to keep properly, but have added some regex to do this.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +121,108 @@ def carry_forwards(df, backdata, impute_vars):
+    for target in target_vars:
+        link_df[f"{target}_link"] = link_df[target] / link_df[f"{target}_prev"]
+
+    # TODO How to deal with infs or 0/0
```
suggestion: could use np.where like in imputation.apportionment.py, or .loc to avoid division by zero.

dpurches replied:
I think this definitely a good shout for clarity, but I don't think we need to do it as we get rid of the NaNs produced by zero division using the `no_zero_mask ` in line 222.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +121,108 @@ def carry_forwards(df, backdata, impute_vars):
+    link_df = (link_df[["reference", "imp_class"] + target_vars + target_vars_prev]
+               .groupby(["reference", "imp_class"])
+               .sum()
+               )
```
looks good. Do you need to do a reset_index()?

dpurches replied:
We already do it on line 160, but I guess it makes more sense to do it here given it's the last operation using the groupby.

dpurches replied:
Have updated this now - weren't selecting the columns to keep properly, but have added some regex to do this.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +121,108 @@ def carry_forwards(df, backdata, impute_vars):
+        group[f"{var}_mor"] = (
+            group.loc[~group[f"{var}_link_trim"], f"{var}_link"].mean()
+        )
+    return group
```
much nicer than that "mean dictionary" we use in earlier code.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +121,108 @@ def carry_forwards(df, backdata, impute_vars):
+    # per imputation class
+    mor_df = (mor_df[["imp_class"] + [f"{var}_mor" for var in target_vars]]
+              .groupby("imp_class").first())
+
```
nice, haven't checked it in the debug, but looks good

AnneONS reviewed (commented):
Looks good so far! 
Main point is to sort out division by zero. 
Also, drop _merge from the dataframe.
It would be nice to see the _prev cols in the qa (can't see them, may be missing something).
Until this is merged, it would be good to either output an additional csv for MoR, or to dump more stuff in the imputation_qa, and for now only output longforms so it doesn't get too big. But for Ana working in DAP, having an extra csv output would be ideal if not too much work.

Fab! look forward to giving it a more thorough go through mid next week.

dpurches committed:
Computed imputed breakdown values and produce links qa output.

dpurches committed:
Merge branch 'RDRP-215' of https://github.com/ONSdigital/research-and-development into RDRP-215

dpurches committed:
Methods updates and renaming.

dpurches committed:
Pre commit updates.

dpurches committed:
Update MoR method to not include nulls in mean calculation.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -19,17 +21,25 @@ def run_mor(df, backdata, impute_vars):
+    )
     # TODO Remove the `XXX_prev` columns (left in for QA)
-    return pd.concat([remainder_df, carried_forwards_df]).reset_index(drop=True)
+    return pd.concat([remainder_df, carried_forwards_df]).reset_index(), links_df
```
I think you need to put that drop=True back in to the reset index, otherwise it creates an extra column called index (and that's even though we have index=False in the write_csv function)

dpurches replied:
Not sure why I got rid of it in the first place! Will do.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -19,17 +21,25 @@ def run_mor(df, backdata, impute_vars):
+    carried_forwards_df = apply_links(
+        carried_forwards_df, links_df, lf_target_vars, config
+    )
     # TODO Remove the `XXX_prev` columns (left in for QA)
```
we can get rid of this, can't we?

dpurches replied:
The #TODO? Yeh we can!

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +125,148 @@ def carry_forwards(df, backdata, impute_vars):
+    for var in target_vars:
+        group = trim_bounds(group, f"{var}_gr", config)
+        # Create mask to not use 0s in mean calculation
+        no_zero_mask = pd.notnull(group[f"{var}_gr"]) & (group[f"{var}_gr"] != 0)
```
A comment in here, perhaps, to explain:

 if there are no non-null, non-zero values in the trimmed group, set the link to 1

dpurches replied:
Good shout.

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +125,148 @@ def carry_forwards(df, backdata, impute_vars):
+                for var in target_vars
+            ]
+        )
+    )
```
nice!

AnneONS commented on src/imputation/MoR.py:
```diff
@@ -112,8 +125,148 @@ def carry_forwards(df, backdata, impute_vars):
+        gr_df[["reference", "imp_class"] + target_vars + target_vars_prev]
+        .groupby(["reference", "imp_class"])
+        .sum()
+    ).reset_index()
```
again, I think you might need drop=True in here

dpurches replied:
I did remove this but realised it broke the code as the index here is "imp_class" which we need back in as a column.

AnneONS replied:
I didn't realise it worked like that. Cool.

AnneONS reviewed (approved):
Looks good! I haven't thoroughly checked the sums in the output, but the code looks good. 

If you could just improve it so we don't get an extra index col in the final df, that would be good.

dpurches commented:
Thanks for the review @AnneONS. Have just spoken to Ana and she is happy with the outputs, bar the fact the emp and HC totals haven't been added yet. Will do that this morning as well as addressing your comments.

dpurches committed:
Address PR comments.

dpurches committed:
Add functionality to calculate totals.

dpurches commented on src/imputation/MoR.py:
```diff
@@ -19,17 +21,26 @@ def run_mor(df, backdata, impute_vars):
+        carried_forwards_df, links_df, lf_target_vars, config
+    )
+    # Calculate totals as with TMI
+    carried_forwards_df = calculate_totals(carried_forwards_df)
```
@AnneONS Have added this line to calculate the totals, but didn't know if we should just call the function once in `imputation_main`?

AnneONS replied:
it was the original plan to do that... but if so, we'd need to do it separately for long forms, where we need to calculate emp_total and headcount_total, and for shortforms, where those 2 are actually target vars, and different totals are needed.

dpurches replied:
Makes sense, are you happy for me to merge?

dpurches committed:
Undo dropping.

dpurches committed:
Merge branch 'develop' of https://github.com/ONSdigital/research-and-development into RDRP-215

dpurches committed:
Update method to use 817 fix and add temporary fix for q201 issue.

dpurches committed:
Update default network_or_hdfs parameter.
